File: write-api-server-kotlin/src/main/kotlin/com/github/j5ik2o/adceet/api/write/domain/AccountId.kt
package com.github.j5ik2o.adceet.api.write.domain

import arrow.core.Either
import wvlet.airframe.ulid.ULID
import java.lang.NumberFormatException

data class AccountId(val value: ULID = ULID.newULID()) : ValueObject {
  companion object {
    fun parseFromString(text: String): Either<Exception, AccountId> {
      return try {
        Either.Right(AccountId(ULID.fromString(text)))
      } catch (ex: Exception) {
        when (ex) {
          is NumberFormatException, is IllegalArgumentException ->
            Either.Left(ex)
          else -> throw ex
        }
      }
    }
  }

  fun asString() = value.toString()
}

File: write-api-server-scala/src/main/scala/com/github/j5ik2o/adceet/api/write/domain/AccountId.scala
package com.github.j5ik2o.adceet.api.write.domain

import wvlet.airframe.ulid.ULID

final case class AccountId(value: ULID = ULID.newULID) extends ValueObject

object AccountId {

  def parseFromString(text: String): Either[Exception, AccountId] = {
    try {
      Right(AccountId(ULID.fromString(text)))
    } catch {
      case ex: NumberFormatException    => Left(ex)
      case ex: IllegalArgumentException => Left(ex)
      case ex: Throwable                => throw ex
    }
  }

}
